feature/#6565 Script Units Expected to Submit Report

EvanSun220 opened the pull request:
### **Related Ticket:**

- ticket [#6565](https://github.com/US-EPA-CAMD/easey-ui/issues/6565)

### **Updates:**

- Updated view camdecmpsaux.vw_em_submission_access
- Added new function camdecmps.get_em_reporting_status
- Added new function get_units_expected_to_submit_report_data to get data for units expected to submit report

EvanSun220 committed:
feature/#6565 update view camdecmpsaux.vw_em_submission_access, add new function camdecmps.get_em_reporting_status, add new function get_units_expected_to_submit_report_data to get data for units expected to submit report

EvanSun220 committed:
feature/#6565 add CASCADE keyword when drop functions, update the logic related with submission_status column in view vw_em_submission_access

mark-hayward-erg commented on deployments/ecmps/release-v2.0-fix/Sprint18/6565/camdecmps/get_em_reporting_status.sql:
```diff
@@ -0,0 +1,173 @@
+DROP FUNCTION IF EXISTS camdecmps.get_em_reporting_status(numeric, numeric, character varying) CASCADE;
```
Deployment scripts (generally speaking) should not include DROP statements. A DROP statement would only be needed in circumstances where the signature of an existing function/procedure is being changed - AND - in that case, scripts for any dependent objects would also need to be included in the deployments folder.

The CREATE OR REPLACE statement is sufficient to create or modify a function, procedure, or view.

mark-hayward-erg commented on deployments/ecmps/release-v2.0-fix/Sprint18/6565/camdecmpsaux/vw_em_submission_access.sql:
```diff
@@ -0,0 +1,114 @@
+-- View: camdecmpsaux.vw_em_submission_access
+
+DROP VIEW IF EXISTS camdecmpsaux.vw_em_submission_access;
```
Deployment scripts (generally speaking) should not include DROP statements. A DROP statement would only be needed in circumstances where the signature of an existing function/procedure is being changed - AND - in that case, scripts for any dependent objects would also need to be included in the deployments folder.

The CREATE OR REPLACE statement is sufficient to create or modify a function, procedure, or view.

mark-hayward-erg commented on deployments/ecmps/release-v2.0-fix/Sprint18/6565/get_units_expected_to_submit_report_data.sql:
```diff
@@ -0,0 +1,87 @@
+DROP FUNCTION IF EXISTS get_units_expected_to_submit_report_data(numeric, character varying, character varying, character varying, numeric, numeric, character varying) CASCADE;
```
Deployment scripts (generally speaking) should not include DROP statements. A DROP statement would only be needed in circumstances where the signature of an existing function/procedure is being changed - AND - in that case, scripts for any dependent objects would also need to be included in the deployments folder.

The CREATE OR REPLACE statement is sufficient to create or modify a function, procedure, or view.

mark-hayward-erg commented on deployments/ecmps/release-v2.0-fix/Sprint18/6565/get_units_expected_to_submit_report_data.sql:
```diff
@@ -0,0 +1,87 @@
+DECLARE
+    v_rpt_period_id numeric;
+BEGIN
+    SELECT RPT_PERIOD_ID
```
need to add logic to verify that the required inputs (program, year, and quarter) have been passed in.  If any of the required values are not passed in, the function should use RAISE NOTICE to inform the caller that these parameters are required.

mark-hayward-erg commented on deployments/ecmps/release-v2.0-fix/Sprint18/6565/get_units_expected_to_submit_report_data.sql:
A script for this function also needs to be added to camdecmpsaux/functions.

mark-hayward-erg commented on deployments/ecmps/release-v2.0-fix/Sprint18/6565/get_units_expected_to_submit_report_data.sql:
Need to include "camdecmpsaux." in the DROP and CREATE OR REPLACE statements.

mark-hayward-erg commented on deployments/ecmps/release-v2.0-fix/Sprint18/6565/get_units_expected_to_submit_report_data.sql:
```diff
@@ -0,0 +1,87 @@
+)
+RETURNS TABLE (
+    
+    ORIS_CODE                     character varying,
```
Type should be numeric

mark-hayward-erg commented on deployments/ecmps/release-v2.0-fix/Sprint18/6565/get_units_expected_to_submit_report_data.sql:
```diff
@@ -0,0 +1,87 @@
+    FACILITY_NAME                 character varying,
+    STATE                         character varying,
+    UNIT_ID                       numeric,
+    LOCATIONS                     character varying,
```
type should be text

mark-hayward-erg commented on deployments/ecmps/release-v2.0-fix/Sprint18/6565/get_units_expected_to_submit_report_data.sql:
```diff
@@ -0,0 +1,87 @@
+    EM_SUB_TYPE_CD_DESCRIPTION    character varying,
+    ACCESS_BEGIN_DATE             date,
+    ACCESS_END_DATE               date,
+    WINDOW_STATUS                 character varying,
```
type should be text

mark-hayward-erg commented on deployments/ecmps/release-v2.0-fix/Sprint18/6565/get_units_expected_to_submit_report_data.sql:
```diff
@@ -0,0 +1,87 @@
+    ACCESS_BEGIN_DATE             date,
+    ACCESS_END_DATE               date,
+    WINDOW_STATUS                 character varying,
+    SUBMISSION_STATUS             character varying,
```
type should be text

mark-hayward-erg commented on deployments/ecmps/release-v2.0-fix/Sprint18/6565/get_units_expected_to_submit_report_data.sql:
```diff
@@ -0,0 +1,87 @@
+    ORIS_CODE                     character varying,
+    FACILITY_NAME                 character varying,
+    STATE                         character varying,
+    UNIT_ID                       numeric,
```
This should be UNITID (not UNIT_ID - this was incorrect in the ticket description, which I have corrected) and type should be character varying

mark-hayward-erg commented on deployments/ecmps/release-v2.0-fix/Sprint18/6565/get_units_expected_to_submit_report_data.sql:
```diff
@@ -0,0 +1,87 @@
+        F.ORIS_CODE,
+        F.FACILITY_NAME,
+        F.STATE,
+        U.UNIT_ID,
```
Should be U.UNITID (this was wrong in the ticket description; I have since corrected this)

mark-hayward-erg commented on deployments/ecmps/release-v2.0-fix/Sprint18/6565/get_units_expected_to_submit_report_data.sql:
```diff
@@ -0,0 +1,87 @@
+             SUBMISSION_ID,
+             SUBMIT_DATE,
+             SEVERITY_CD_DESCRIPTION
+         FROM camdecmpsaux.VW_EM_SUBMISSION_ACCESS
```
need to alias camdecmpsaux.VW_EM_SUBMISSION_ACCESS (as esa_sub) and include "esa_sub." for all columns selected.

mark-hayward-erg commented on deployments/ecmps/release-v2.0-fix/Sprint18/6565/get_units_expected_to_submit_report_data.sql:
```diff
@@ -0,0 +1,87 @@
+             WINDOW_STATUS,
+             SUBMISSION_STATUS,
+             SUBMISSION_ID,
+             SUBMIT_DATE,
```
should be submission_date

mark-hayward-erg commented on deployments/ecmps/release-v2.0-fix/Sprint18/6565/get_units_expected_to_submit_report_data.sql:
```diff
@@ -0,0 +1,87 @@
+    WINDOW_STATUS                 character varying,
+    SUBMISSION_STATUS             character varying,
+    SUBMISSION_ID                 bigint,
+    SUBMIT_DATE                   timestamp,
```
should be SUBMISSION_DATE

mark-hayward-erg commented on deployments/ecmps/release-v2.0-fix/Sprint18/6565/get_units_expected_to_submit_report_data.sql:
```diff
@@ -0,0 +1,87 @@
+        COALESCE(ESA.WINDOW_STATUS, 'No Window') AS WINDOW_STATUS,
+        ESA.SUBMISSION_STATUS,
+        ESA.SUBMISSION_ID,
+        ESA.SUBMIT_DATE,
```
should be SUBMISSION_DATE

mark-hayward-erg commented on deployments/ecmps/release-v2.0-fix/Sprint18/6565/get_units_expected_to_submit_report_data.sql:
```diff
@@ -0,0 +1,87 @@
+    FROM
+        (
+         SELECT 
+             UNIT_ID,
```
Need to include UNITID column

mark-hayward-erg commented on deployments/ecmps/release-v2.0-fix/Sprint18/6565/get_units_expected_to_submit_report_data.sql:
```diff
@@ -0,0 +1,87 @@
+             UNIT_ID,
+             FAC_ID,
+             camdecmps.get_em_reporting_status(UNIT_ID, v_rpt_period_id, V_PRG_CODE) AS MON_PLAN_ID
+         FROM camd.UNIT
```
need to alias CAMD.UNIT (as u_sub) and add "u_sub." to selected columns

mark-hayward-erg commented on deployments/ecmps/release-v2.0-fix/Sprint18/6565/camdecmps/get_em_reporting_status.sql:
```diff
@@ -0,0 +1,173 @@
+            SELECT ue.up_id
+              FROM camd.unit_exemption ue
+              INNER JOIN camd.unit_program up2 ON ue.unit_id = up2.unit_id
+              INNER JOIN camd.program_exemption pe ON up2.prg_cd = pe.prg_cd 
```
should be camdmd.program_exemption

mark-hayward-erg commented on camdecmps/functions/get_em_reporting_status.sql:
```diff
@@ -0,0 +1,173 @@
+            SELECT ue.up_id
+              FROM camd.unit_exemption ue
+              INNER JOIN camd.unit_program up2 ON ue.unit_id = up2.unit_id
+              INNER JOIN camd.program_exemption pe ON up2.prg_cd = pe.prg_cd 
```
should be should be camdmd.program_exemption

mark-hayward-erg reviewed (changes requested):
Also need to resolve conflicts on vw_em_submission_access (to accomodate changes made for #106  that are now in develop).

EvanSun220 committed:
feature/#6565 solve the merge conflict with develop branch, update functions camdecmpsaux.get_units_expected_to_submit_report_data and camdecmps.get_em_reporting_status

EvanSun220 committed:
Merge branch 'develop' into feature/#6565-Script-Units-Expected-to-Submit-Report

mark-hayward-erg committed:
refactoring for performance; remove use of functions within views; replace needed functions with view equivalents; add additional indexes; includes functional updates for the submission window job

mark-hayward-erg committed:
update index scripts

mark-hayward-erg committed:
Merge branch 'develop' of https://github.com/US-EPA-CAMD/easey-db-scripts into feature/#6565-Script-Units-Expected-to-Submit-Report

# Conflicts resolved:
#	camdecmpsaux/functions/get_last_submission.sql (deleted in this branch)
#	camdecmpsaux/procedures/init_and_close_em_submission_access.sql (use version from this branch)
#	camdecmpsaux/views/vw_submission_list.sql (use version from this branch)

mark-hayward-erg reviewed (dismissed):
Corrected identified issues as part of performance optimization.

mark-hayward-erg committed:
restore camdecmpswks/functions/get_mp_location_list.sql (function is used directly by APIs for a single MP at a time)

mark-hayward-erg committed:
repurpose camdecmps/procedures/initialize_em_submission_access.sql as a misc script (used to recreate em windows for a specified year and quarter)
update init_and_close_em_submission_access.sql to use v_sysdate_as_date (not current_date) when setting window dates (procedure can be executed to act like any date)
remove camdecmpswks.get_mp_location_list from drop_functions.sql (function is needed and used directly by APIs)
